File: source/VariantAttribute.m
classdef VariantAttribute
    %VARIANTATTRIBUTE A class representing the attribute for a given
    %variant of a word.
    %
    % Syntax:
    %   A = VariantAttribute(V, IS)
    %       Create an attribute for the current variant that states:
    %       - it is attested for the category V
    %       - is is/it is not the reference for the current category.

    % Copyright 2023 Acadèmia de su Sardu APS

    properties(SetAccess=immutable)
        % The variant's category
        Category

        % True if the variant is the reference the current given category.
        IsCategoryReference
    end

    methods
        function obj = VariantAttribute(category, isCategoryReference)
            %VARIANTATTRIBUTE Construct the object
            arguments
                category (1,1) {mustBeTextScalar}
                isCategoryReference (1,1) logical
            end

            category = string(category);
            validateCategory(category);
            obj.Category = category;
            obj.IsCategoryReference = isCategoryReference;
        end
    end
end

File: python/linguistic_variation_toolbox/variant_attribute.py
from __future__ import annotations
from ._private.validate_category import validate_category


class VariantAttribute:
    """Attribute for a given variant: its category and whether it is the reference."""

    def __init__(self, category: str, is_category_reference: bool) -> None:
        validate_category(category)
        self._category = str(category)
        self._is_category_reference = bool(is_category_reference)

    @property
    def category(self) -> str:
        return self._category

    @property
    def is_category_reference(self) -> bool:
        return self._is_category_reference

    def __repr__(self) -> str:
        return (
            f"VariantAttribute(category={self._category!r}, "
            f"is_category_reference={self._is_category_reference})"
        )
